src/metabase/models/data_permissions.clj

```diff
@@ -219,93 +218,103 @@
+                      {})))
+    (t2/with-transaction [_conn]
+      (let [group-id               (u/the-id group-or-id)
+            new-perms              (map (fn [[table value]]
+                                          (let [{:keys [id db_id schema]}
+                                                (if (map? table)
+                                                  table
+                                                  (t2/select-one [:model/Table :id :db_id :schema] :id table))]
+                                            {:type       perm-type
+                                             :group_id   group-id
+                                             :perm_value value
+                                             :db_id      db_id
+                                             :table_id   id
+                                             :schema     schema}))
+                                        table-perms)
+            _                      (when (not= (count (set (map :db_id new-perms))) 1)
+                                     (throw (ex-info (tru "All tables must belong to the same database.")
+                                                     {:new-perms new-perms})))
+            table-ids              (map :table_id new-perms)
+            db-id                  (:db_id (first new-perms))
+            existing-db-perm       (t2/select-one :model/DataPermissions
+                                                  {:where
+                                                   [:and
+                                                    [:= :type (name perm-type)]
+                                                    [:= :group_id group-id]
+                                                    [:= :db_id db-id]
+                                                    [:= :table_id nil]]})
+            existing-db-perm-value (:perm_value existing-db-perm)]
+        (if existing-db-perm
+          (when (not= values #{existing-db-perm-value})
+            ;; If we're setting any table permissions to a value that is different from the database-level permission,
+            ;; we need to replace it with individual permission rows for every table in the database instead.
+            (let [other-tables    (t2/select :model/Table {:where [:and
+                                                                   [:= :db_id db-id]
+                                                                   [:not [:in :id table-ids]]]})
+                  other-new-perms (map (fn [table]
+                                         {:type       perm-type
+                                          :group_id   group-id
+                                          :perm_value existing-db-perm-value
+                                          :db_id      db-id
+                                          :table_id   (:id table)
+                                          :schema     (:schema table)})
```

Comment:
I don't think we need to — the table should store the same thing for the schema field as what's in the `metabase_table` table. Coercing `nil` schemas to empty strings is just a presentational thing when fetching the graph, so that it's consistent with how the existing graph works